sklearn/utils/__init__.py

```diff
@@ -186,7 +186,11 @@ def _pandas_indexing(X, key, key_dtype, axis):
         key = list(key)
     # check whether we should index with loc or iloc
     indexer = X.iloc if key_dtype == 'int' else X.loc
-    return indexer[:, key] if axis else indexer[key]
+    sliced_X = indexer[:, key] if axis else indexer[key]
+    # quick hack to silent redundant pandas SettingWithCopyWarning
+    if hasattr(sliced_X, "_is_copy"):
+        sliced_X._is_copy = None
```

Comment:
I would not remove this here. We might silent too much true positive. We need to move this part inside the cross-validation. We cannot be sure that people do the write thing in the other case.

Reply:
I don't think we'd be silencing true positives. safe_indexing applies where numpy fancy indexing is otherwise used. Copying is normal semantics and should be assumed intentional in any application of safe_indexing. The docstring can be updated if this is not clear enough

Reply:
There is one case where `safe_indexing` is not doing a fancy indexing, which is to select a feature with `axis=1` (used in eg column transformer and partial dependence). 

Now, I think that for such a case, there won't be settingwithcopy warnings. So I assume setting the attribute to None won't be problematic in that case. 
For the rest I think @jnothman is correct, indexing with a list-like on the rows is fancy indexing and always gives a copy, so we never want the warning.

(given the above, it might be cleaner to have separate functions for selecting rows vs selecting features)

Reply:
I answered @glemaitre request by moving my fix in the cross validation function. Let me know if you prefer the more general solution in safe_indexing. I would revert my last commit.